### PR TITLE
Include test and validation process in LSTM.ipynb

### PR DESCRIPTION
**Pull request by bradyhhhuang**
add following things in LSTM.ipynb
1. add process_data function in AudioLSTM class for the forward process of validation and test
2. add the validation process in epoch training
3. add the test process and saving predict results

Note that the loss between y_test_pred and y_test_target seems large although I feel fine when listening the output file.
Maybe we should increase epochs or hidden_size?

Also, please help to double confirm the sheet [here](https://docs.google.com/spreadsheets/d/1uHMal6m7UF-ALVxWaDk3HJlecuI3nfttT1cZ-Aw3ZnE/edit?gid=0#gid=0) for the important metrics we should record. I will follow the sheet to record the metric we need after our confirmation.

### CODE REVIEW

File: GuitarLSTM/LSTM.ipynb
Comment (bytehut) [inline GuitarLSTM/LSTM.ipynb:1]: whats the reasoning behind detaching and resetting the hidden state every so often besides memory reasons? seems like it would cause issues in our predicted output

File: GuitarLSTM/LSTM.ipynb [hunk 1]
```diff
@@ -1,235 +1 @@
...
..., :, 0][:1000]\n","plt.plot(range(len(y_test_pred)), y_test_pred, label='pred')\n","plt.plot(range(len(y_test_true)), y_test_true, label='target')\n","plt.legend()"]},{"cell_type":"code","execution_count":24,"metadata":{"executionInfo":{"elapsed":423,"status":"ok","timestamp":1732694136418,"user":{"displayName":"Hua-Hsing Huang","userId":"01596834180958819638"},"user_tz":480},"id":"os6Sy-TrL4Eb"},"outputs":[],"source":["import torchaudio\n","\n","filepath = \"result/ht1_target_predict.wav\"\n","output1 = output[0]\n","torchaudio.save(filepath, output1, sample_rate=44100, channels_first=False)"]},{"cell_type":"code","execution_count":14,"metadata":{"executionInfo":{"elapsed":4,"status":"ok","timestamp":1732693779882,"user":{"displayName":"Hua-Hsing Huang","userId":"01596834180958819638"},"user_tz":480},"id":"C5oGZ6d5eedt"},"outputs":[],"source":[]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.11.9"}},"nbformat":4,"nbformat_minor":0}
```
Comment (bytehut) [inline GuitarLSTM/LSTM.ipynb:1]: When creating the graph for comparison between predicted and target I noticed you took only one channel and compared it rather than both channels, this could possibly be our issue? very strange how our predicted output takes the form of a big sine wave though

**Review by bytehut — COMMENTED** [review]
Good work, thanks for finishing up the LSTM module.

I have concerns around training with a short sequence length, but validating and testing with a much longer sequence length. We can experiment with tweaking parameters to hopefully get an even better performing LSTM but this looks great!

**Review by bytehut — APPROVED** [review]

**Review by mxux — APPROVED** [review]
LGTM
